File: youtube/lat8-heart/try1.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt 
import seaborn as sns 
import pandas as pd
import matplotlib.pyplot as plt 
import seaborn as sns 
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, classification_report
from sklearn.metrics import confusion_matrix
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.neural_network import MLPClassifier
from sklearn.neural_network import MLPRegressor
from sklearn.preprocessing import MinMaxScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.neural_network import MLPClassifier
from sklearn.neural_network import MLPRegressor

heart = pd.read_csv('ML/youtube/lat8-heart/heart.csv')
print(heart)
print(heart.info())
print(heart.describe())
print(heart.isnull().sum())

for col in heart:
    print(col)
    print(heart[col].unique())
    print('\n\n')
    
# fig, ax = plt.subplots(3, 5, figsize=(15, 10))
# sns.scatterplot(x=heart.index, y=heart['age'], ax=ax[0, 0], color='blue')
# sns.scatterplot(x=heart.index, y=heart['sex'], ax=ax[0, 1], color='green')
# sns.scatterplot(x=heart.index, y=heart['cp'], ax=ax[0, 2], color='orange')
# sns.scatterplot(x=heart.index, y=heart['trestbps'], ax=ax[0, 3], color='red')
# sns.scatterplot(x=heart.index, y=heart['chol'], ax=ax[0, 4], color='blue')
# sns.scatterplot(x=heart.index, y=heart['fbs'], ax=ax[1, 0], color='green')
# sns.scatterplot(x=heart.index, y=heart['restecg'], ax=ax[1, 1], color='orange')
# sns.scatterplot(x=heart.index, y=heart['thalach'], ax=ax[1, 2], color='red')
# sns.scatterplot(x=heart.index, y=heart['exang'], ax=ax[1, 3], color='blue')
# sns.scatterplot(x=heart.index, y=heart['oldpeak'], ax=ax[1, 4], color='green')
# sns.scatterplot(x=heart.index, y=heart['slope'], ax=ax[2, 0], color='orange')
# sns.scatterplot(x=heart.index, y=heart['ca'], ax=ax[2, 1], color='red')
# sns.scatterplot(x=heart.index, y=heart['thal'], ax=ax[2, 2], color='red')
# sns.scatterplot(x=heart.index, y=heart['target'], ax=ax[2, 3], color='red')
# plt.tight_layout()
# plt.show()

q1 = heart['age'].quantile(0.25)
q3 = heart['age'].quantile(0.75)
iqr = q3 - q1
heart = heart[(heart['age'] >= q1 - 1.5*iqr) & (heart['age'] <= q3 + 1.5*iqr )]

q1 = np.percentile(heart['sex'], 25)
q3 = np.percentile(heart['sex'], 75)
iqr = q3 - q1
heart = heart[(heart['sex'] >= q1 - 1.5*iqr) & (heart['sex'] <= q3 + 1.5*iqr )]

q1 = heart['cp'].quantile(0.25)
q3 = heart['cp'].quantile(0.75)
iqr = q3 - q1
heart = heart[(heart['cp'] >= q1 - 1.5*iqr) & (heart['cp'] <= q3 + 1.5*iqr )]

q1 = np.percentile(heart['trestbps'], 25)
q3 = np.percentile(heart['trestbps'], 75)
iqr = q3 - q1
heart = heart[(heart['trestbps'] >= q1 - 1.5*iqr) & (heart['trestbps'] <= q3 + 1.5*iqr )]

q1 = heart['chol'].quantile(0.25)
q3 = heart['chol'].quantile(0.75)
iqr = q3 - q1
heart = heart[(heart['chol'] >= q1 - 1.5*iqr) & (heart['chol'] <= q3 + 1.5*iqr )]

q1 = np.percentile(heart['fbs'], 25)
q3 = np.percentile(heart['fbs'], 75)
iqr = q3 - q1
heart = heart[(heart['fbs'] >= q1 - 1.5*iqr) & (heart['fbs'] <= q3 + 1.5*iqr )]

q1 = heart['restecg'].quantile(0.25)
q3 = heart['restecg'].quantile(0.75)
iqr = q3 - q1
heart = heart[(heart['restecg'] >= q1 - 1.5*iqr) & (heart['restecg'] <= q3 + 1.5*iqr )]

q1 = np.percentile(heart['thalach'], 25)
q3 = np.percentile(heart['thalach'], 75)
iqr = q3 - q1
heart = heart[(heart['thalach'] >= q1 - 1.5*iqr) & (heart['thalach'] <= q3 + 1.5*iqr )]

q1 = heart['exang'].quantile(0.25)
q3 = heart['exang'].quantile(0.75)
iqr = q3 - q1
heart = heart[(heart['exang'] >= q1 - 1.5*iqr) & (heart['exang'] <= q3 + 1.5*iqr )]

q1 = np.percentile(heart['oldpeak'], 25)
q3 = np.percentile(heart['oldpeak'], 75)
iqr = q3 - q1
heart = heart[(heart['oldpeak'] >= q1 - 1.5*iqr) & (heart['oldpeak'] <= q3 + 1.5*iqr )]

q1 = heart['slope'].quantile(0.25)
q3 = heart['slope'].quantile(0.75)
iqr = q3 - q1
heart = heart[(heart['slope'] >= q1 - 1.5*iqr) & (heart['slope'] <= q3 + 1.5*iqr )]

q1 = np.percentile(heart['ca'], 25)
q3 = np.percentile(heart['ca'], 75)
iqr = q3 - q1
heart = heart[(heart['ca'] >= q1 - 1.5*iqr) & (heart['ca'] <= q3 + 1.5*iqr )]

q1 = np.percentile(heart['thal'], 25)
q3 = np.percentile(heart['thal'], 75)
iqr = q3 - q1
heart = heart[(heart['thal'] >= q1 - 1.5*iqr) & (heart['thal'] <= q3 + 1.5*iqr )]

q1 = np.percentile(heart['target'], 25)
q3 = np.percentile(heart['target'], 75)
iqr = q3 - q1
heart = heart[(heart['target'] >= q1 - 1.5*iqr) & (heart['target'] <= q3 + 1.5*iqr )]

# fig, ax = plt.subplots(3, 5, figsize=(15, 10))
# sns.scatterplot(x=heart.index, y=heart['age'], ax=ax[0, 0], color='blue')
# sns.scatterplot(x=heart.index, y=heart['sex'], ax=ax[0, 1], color='green')
# sns.scatterplot(x=heart.index, y=heart['cp'], ax=ax[0, 2], color='orange')
# sns.scatterplot(x=heart.index, y=heart['trestbps'], ax=ax[0, 3], color='red')
# sns.scatterplot(x=heart.index, y=heart['chol'], ax=ax[0, 4], color='blue')
# sns.scatterplot(x=heart.index, y=heart['fbs'], ax=ax[1, 0], color='green')
# sns.scatterplot(x=heart.index, y=heart['restecg'], ax=ax[1, 1], color='orange')
# sns.scatterplot(x=heart.index, y=heart['thalach'], ax=ax[1, 2], color='red')
# sns.scatterplot(x=heart.index, y=heart['exang'], ax=ax[1, 3], color='blue')
# sns.scatterplot(x=heart.index, y=heart['oldpeak'], ax=ax[1, 4], color='green')
# sns.scatterplot(x=heart.index, y=heart['slope'], ax=ax[2, 0], color='orange')
# sns.scatterplot(x=heart.index, y=heart['ca'], ax=ax[2, 1], color='red')
# sns.scatterplot(x=heart.index, y=heart['thal'], ax=ax[2, 2], color='red')
# sns.scatterplot(x=heart.index, y=heart['target'], ax=ax[2, 3], color='red')
# plt.tight_layout()
# plt.show()

numvars = ['age', 'trestbps', 'chol', 'thalach', 'oldpeak']
scaler = MinMaxScaler()
heart[numvars] = scaler.fit_transform(heart[numvars])

print(heart)

x = heart
y = heart.pop('target')


X_train, X_test, y_train, y_test = train_test_split(x,y, test_size=0.3, random_state=42, stratify=y)
print(y_train.value_counts(normalize = True))
print(y_test.value_counts(normalize = True))


k_folds = KFold(n_splits=5, shuffle=True, random_state=42)
modelLR  = LogisticRegression()
scoring = 'accuracy'
score = cross_val_score(modelLR, X_train, y_train, cv = k_folds, n_jobs=1, scoring=scoring)
print(score)
print(round(score.mean(), 2))



modelRF  = RandomForestClassifier()
scoring = 'accuracy'
score = cross_val_score(modelRF, X_train, y_train, cv = k_folds, n_jobs=1, scoring=scoring)
print(score)
print(round(score.mean(), 2))



modelMPL  = MLPClassifier()
scoring = 'accuracy'
score = cross_val_score(modelMPL, X_train, y_train, cv = k_folds, n_jobs=1, scoring=scoring)
print(score)
print(round(score.mean(), 2))


modelNB  = GaussianNB()
scoring = 'accuracy'
score = cross_val_score(modelNB, X_train, y_train, cv = k_folds, n_jobs=1, scoring=scoring)
print(score)
print(round(score.mean(), 2))

modelRF.fit(X_train, y_train)
y_pred_train = modelRF.predict(X_train)

train = pd.DataFrame({
    'y_train': y_train,
    'y_pred_train': y_pred_train
})
print(train)
print(train[train['y_train'] != train['y_pred_train']])

akurasi = accuracy_score(y_train, y_pred_train)
print(akurasi)
cm = confusion_matrix(y_train, y_pred_train)
print(cm)
sns.heatmap(cm, annot=True,  fmt='.0f', cmap=plt.cm.Blues)
plt.xlabel('y_pred_train')
plt.ylabel('y_train')
plt.title('confusion matrix')
plt.show()




y_pred_test = modelRF.predict(X_test)

train = pd.DataFrame({
    'y_test': y_test,
    'y_pred_test': y_pred_test
})
print(train)
print(train[train['y_test'] != train['y_pred_test']])

akurasi = accuracy_score(y_test, y_pred_test)
print(akurasi)
cm = confusion_matrix(y_test, y_pred_test)
print(cm)
sns.heatmap(cm, annot=True,  fmt='.0f', cmap=plt.cm.Blues)
plt.xlabel('y_pred_test')
plt.ylabel('y_test')
plt.title('confusion matrix')
plt.show()

File: youtube/lat8-heart/try1.r
library(dplyr)
library(ggplot2)
library(stringr)
library(gridExtra)
library(RMySQL)
library(dplyr)
library(caTools)
library(tidyr)
library(plyr)
library(glmnet)
library(lubridate)
library(tidyverse)
library(randomForest)
library(caret)
library(patchwork)
library(nnet)
library(klaR) # untuk Naive Bayes
library(rpart) # untuk Decision Trees



df <- read.csv("ML/youtube/lat8-heart/heart.csv")
print(head(df))
print(dim(df))
print(str(df))
print(summary(df))
print(colSums(is.na(df)))
print(sapply(df, class))
print(sapply(df, function(x) length(unique(x[!is.na(x)]))))


## TIDAK BISA KARENA ADANYA TABRAKAN DENGAN PACKAGE LAINNTA
# library(dplyr)
# library(tidyverse)
# library(ggplot2)
# null1 <- df %>% summarise(across(everything(), ~n_distinct(na.omit(.))))
# print(as.data.frame(null1))


for (col in colnames(df)){
    print(col)
    print(unique(df[[col]]))
    cat('\n\n')
}


p1 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = age), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p2 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = sex), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p3 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = cp), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p4 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = trestbps), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal() 
p5 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = chol), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p6 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = fbs), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal() 
p7 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = restecg), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p8 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = thalach), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal() 
p9 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = exang), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p10 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = oldpeak), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p11 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = slope), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p12 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = ca), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p13 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = thal), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p14 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = target), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()

# # Menggabungkan plot menggunakan patchwork
plot <- p1+p2+p3+p4+p5+p6+p7+p8+p9+p10+p11+p12+p13+p14
# plot <- (p1 | p2) / (p3 | p4)
print(plot)



q1 <- quantile(df$age, 0.25)
q3 <- quantile(df$age, 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, age >= q1 - 1.5 * iqr & age <= q3 + 1.5 * iqr)

q1 <- quantile(df$sex, 0.25)
q3 <- quantile(df$sex, 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, sex >= q1 - 1.5 * iqr & sex <= q3 + 1.5 * iqr)

q1 <- quantile(df$cp , 0.25)
q3 <- quantile(df$cp , 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, cp  >= q1 - 1.5 * iqr & cp  <= q3 + 1.5 * iqr)

q1 <- quantile(df$trestbps , 0.25)
q3 <- quantile(df$trestbps , 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, trestbps  >= q1 - 1.5 * iqr & trestbps  <= q3 + 1.5 * iqr)

q1 <- quantile(df$chol  , 0.25)
q3 <- quantile(df$chol  , 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, chol   >= q1 - 1.5 * iqr & chol   <= q3 + 1.5 * iqr)

q1 <- quantile(df$fbs , 0.25)
q3 <- quantile(df$fbs , 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, fbs  >= q1 - 1.5 * iqr & fbs  <= q3 + 1.5 * iqr)

q1 <- quantile(df$restecg , 0.25)
q3 <- quantile(df$restecg , 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, restecg  >= q1 - 1.5 * iqr & restecg  <= q3 + 1.5 * iqr)

q1 <- quantile(df$thalach, 0.25)
q3 <- quantile(df$thalach, 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, thalach >= q1 - 1.5 * iqr & thalach <= q3 + 1.5 * iqr)

q1 <- quantile(df$exang , 0.25)
q3 <- quantile(df$exang , 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, exang  >= q1 - 1.5 * iqr & exang  <= q3 + 1.5 * iqr)

q1 <- quantile(df$oldpeak, 0.25)
q3 <- quantile(df$oldpeak, 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, oldpeak >= q1 - 1.5 * iqr & oldpeak <= q3 + 1.5 * iqr)

q1 <- quantile(df$slope , 0.25)
q3 <- quantile(df$slope , 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, slope  >= q1 - 1.5 * iqr & slope  <= q3 + 1.5 * iqr)

q1 <- quantile(df$ca, 0.25)
q3 <- quantile(df$ca, 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, ca >= q1 - 1.5 * iqr & ca <= q3 + 1.5 * iqr)

q1 <- quantile(df$thal , 0.25)
q3 <- quantile(df$thal , 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, thal  >= q1 - 1.5 * iqr & thal  <= q3 + 1.5 * iqr)

q1 <- quantile(df$target , 0.25)
q3 <- quantile(df$target , 0.75)
iqr <- q3 - q1
# Mengaplikasikan aturan IQR untuk menyaring data
df <- subset(df, target  >= q1 - 1.5 * iqr & target  <= q3 + 1.5 * iqr)



p1 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = age), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p2 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = sex), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p3 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = cp), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p4 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = trestbps), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal() 
p5 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = chol), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p6 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = fbs), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal() 
p7 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = restecg), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p8 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = thalach), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal() 
p9 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = exang), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p10 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = oldpeak), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p11 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = slope), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p12 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = ca), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p13 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = thal), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p14 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = target), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()

# # Menggabungkan plot menggunakan patchwork
plot <- p1+p2+p3+p4+p5+p6+p7+p8+p9+p10+p11+p12+p13+p14
# plot <- (p1 | p2) / (p3 | p4)
print(plot)


print(head(df))

numvars <- c('age', 'trestbps', 'chol', 'thalach', 'oldpeak')
scaler <- preProcess(df[numvars], method = "range")
df[numvars] <- predict(scaler, df[numvars])
print(head(df))

x <- dplyr::select(df, -target)
y <- as.factor(df$target)


set.seed(42)
# index <- createDataPartition(y, p = 0.7, list = TRUE)
# print(head(index))


# x_train <- x[index$Resample1, ]
# x_test <- x[-index$Resample1, ]
# y_train <- y[index$Resample1]
# y_test <- y[-index$Resample1]


split <- sample.split(y, SplitRatio = 0.7)
# x_train <- x[split, ]
# x_test <- x[!split, ]
# y_train <- y[split]
# y_test <- y[!split]

# Membuat data pelatihan
x_train <- subset(x, split == TRUE)
y_train <- subset(y, split == TRUE)
# Membuat data pengujian
x_test <- subset(x, split == FALSE)
y_test <- subset(y, split == FALSE)



ctrl <- trainControl(method="cv", number=5)
modelrf <- train(x_train, y_train, method="rf", trControl=ctrl)
print('accuracy mean')
print(modelrf$result)
print(mean(modelrf$result$Accuracy))
cat('\n')

model <- train(x_train, y_train, method="nnet", trControl=ctrl)
print('accuracy mean')
print(model$result)
print(mean(model$result$Accuracy))
cat('\n')

model <- train(x_train, y_train, method="rpart", trControl=ctrl)
print('accuracy mean')
print(model$result)
print(mean(model$result$Accuracy))
cat('\n')

model <- train(x_train, y_train, method="nb", trControl=ctrl)
print('accuracy mean')
print(model$result)
print(mean(model$result$Accuracy))
cat('\n')



y_pred_train <- predict(modelrf, x_train)
df_pred <- data.frame(
    y_train = y_train,
    y_pred_train = y_pred_train
)
print(head(df_pred))

df_pred <- filter(df_pred, y_train != y_pred_train)
print(head(df_pred))

# Calculate accuracy
accuracy <- confusionMatrix(data = y_pred_train, reference = y_train)$overall['Accuracy']
print(accuracy)

cat('\n\n\n\n')
# Calculate confusion matrix
print('Confusion Matrix')
cm <- confusionMatrix(data = y_pred_train, reference = y_train)
print(cm)




y_pred_test <- predict(modelrf, x_test)
df_pred <- data.frame(
    y_test = y_test,
    y_pred_test = y_pred_test
)
print(head(df_pred))

df_pred <- filter(df_pred, y_test != y_pred_test)
print(head(df_pred))

# Calculate accuracy
accuracy <- confusionMatrix(data = y_pred_test, reference = y_test)$overall['Accuracy']
print(accuracy)

cat('\n\n\n\n')
# Calculate confusion matrix
print('Confusion Matrix')
cm <- confusionMatrix(data = y_pred_test, reference = y_test)
print(cm)
